library/core/src/num/mod.rs

```diff
@@ -846,37 +870,70 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if radix <= 16
+        && src.len()
+            <= mem::size_of::<T>() * 2 - (is_signed_ty && is_positive) as usize
+    {
+        // SAFETY: Consider the highest radix of 16:
+        // u8 max out by ff which is 2 characters, u16 max out by ffff which is 4 characters
+        // We can be sure that any src len of 2 would fit in a u8 so we don't need
+        // to check for overflow.
+        unsafe {
+            if is_positive {
+                // The number is positive
+                for &c in digits {
+                    let x = match (c as char).to_digit(radix) {
+                        Some(x) => x,
+                        None => return Err(PIE { kind: InvalidDigit }),
+                    };
+                    result = result.unchecked_mul(radix);
+                    result = result.unchecked_add(x);
+                }
```

Comment:
The only difference between the positive and the negative branch is the call to `unchecked_add` vs. `unchecked_sub`. Could that be factored out into a private function? Something like
```rust
fn parse_unchecked(
    digits: &[u8],
    radix: u32,
    f: impl Fn(T, u32) -> T
) -> Result<T, PIE>
/* ... */
if is_positive {
    result = parse_unchecked(digits, radix, T::unchecked_add)?;
} else {
    result = parse_unchecked(digits, radix, T::unchecked_sub)?;
}
```

Reply:
Just wondering, would it be faster to parse all numbers as positive and then flip it to negative later if it's negative? Would that be slower?

Reply:
I benched ok_or_else and it made no difference over the match so will tidy up. (Presumably it did make a difference in the past but rustc/llvmcodegen got better).

Also I did notice a slight speedup (2-3%) if we calc mul * radix earlier at the top of the loop as then it has a bit more time to be computed before it's used.

Reply:
> Also I did notice a slight speedup (2-3%) if we calc mul * radix earlier at the top of the loop as then it has a bit more time to be computed before it's used.

Such instruction reordering should already be handled by the compiler. It might be that you're optimizing for your CPU while slowing down other hardware. Do you still see these improvements if you compile for your specific processor with `RUSTFLAGS="-C target-cpu=native"`?

Reply:
It's a noticeble difference with cpu=native too. I'm going to invoke Agner Fog here ( https://agner.org/optimize/instruction_tables.pdf ). Even for the most complex cpus the `IMUL` latency is at least 3 cycles and can be much longer. I would _love_ the compiler to start doing the multiplication earlier but impirical evidence is that it's not (based on one machine). A couple of percent isn't a big deal but people generally have a lot of numbers they want parsing which is why I've been hanging out in this domain for a while. Take a look now I've coded it up - I can switch the muls back to where they were if need be.

Reply:
So, in the original there were two branches that got optimized independently of each other, which hopefully results in optimal performance. In the current version you have a function pointer and `unchecked_{add/sub}` don't get inlined. The reason I suggested a generic function earlier was that it results in full inlining and independent optimization, almost identical to the original.

See https://rust.godbolt.org/z/c6bxGoEWo
\# 1 is the original
\# 2 is the current version
\# 3 is my proposal

What I don't know is if there is any measurable performance difference in this case. Could you check that?

It may or may not make sense to generalize the function to be callable form the checked branch as well, but that's something I'll think about tomorrow.

Reply:
I can't really detect any noticable perf difference between 2 and 3  using long or short numbers. 2 has less instructions but maybe they would be same size if the compiler can figure out that it can inline the add/sub function but it looks to me like it can't as there's a `core::num::<impl u64>::unchecked_add:` in 3's machine code. Am I reading that right?

Reply:
`core::num::<impl u64>::unchecked_add` is in 2's (not 3's) machine code, otherwise yes. Since there doesn't appear to be a perf difference, I retract my function proposal. Trying to extend it to cover the checked case introduces too much boilerplate, so that's not a great path either.

Thanks for checking.